src/utils/leak.js

```diff
@@ -586,7 +586,337 @@ export class LeakImplementation {
+  materializeObject(realm: Realm, val: ObjectValue): void {
     materializeObject(realm, val);
   }
+
+  // This routine materializes objects reachable from non-local bindings read
+  // by a function. It does this for the purpose of outlining calls to that function.
+  //
+  // Notes:
+  // - Locations that are only read need not materialize because their values are up-to-date
+  // at optimization time,
+  // - Locations that are written to are ignored, because we make the assumption, for now,
+  // that the function being outlined is pure.
+  // - Previously havoced locations (#2446) should be reloaded, but are currently rejected.
+  // - Specialization depends on the assumption that the Array op will only be used once.
+  // First, we will enforce it: #2448. Later we will relax it: #2454
+  materializeObjectsTransitive(realm: Realm, outlinedFunction: FunctionValue): void {
+    invariant(realm.isInPureScope());
+    let objectsToMaterialize: Set<ObjectValue> = new Set();
+    let visitedValues: Set<Value> = new Set();
+    computeFromFunctionValue(outlinedFunction);
+
+    if (objectsToMaterialize.size !== 0 && realm.instantRender.enabled) {
+      let error = new CompilerDiagnostic(
+        "Instant Render does not support array operators that reference objects via non-local bindings",
+        outlinedFunction.expressionLocation,
+        "PP0042",
+        "FatalError"
+      );
+      realm.handleError(error);
+      throw new FatalError();
+    }
+
+    for (let object of objectsToMaterialize) {
+      materializeObject(realm, object);
+    }
+
+    return;
+    function computeFromBindings(func: FunctionValue, nonLocalReadBindings: Set<string>): void {
+      invariant(func instanceof ECMAScriptSourceFunctionValue);
+      let environment = func.$Environment;
+      while (environment) {
+        let record = environment.environmentRecord;
+        if (record instanceof ObjectEnvironmentRecord) computeFromValue(record.object);
+        else if (record instanceof DeclarativeEnvironmentRecord || record instanceof FunctionEnvironmentRecord)
+          computeFromDeclarativeEnvironmentRecord(record, nonLocalReadBindings);
+        else if (record instanceof GlobalEnvironmentRecord) break;
+        environment = environment.parent;
+      }
+    }
+    function computeFromDeclarativeEnvironmentRecord(
+      record: DeclarativeEnvironmentRecord,
+      nonLocalReadBindings: Set<string>
+    ): void {
+      let environmentBindings = record.bindings;
+      for (let bindingName of Object.keys(environmentBindings)) {
+        let binding = environmentBindings[bindingName];
+        invariant(binding !== undefined);
+        let found = nonLocalReadBindings.delete(bindingName);
+        // Check what undefined could mean here, besides absent binding
+        // #2446
+        if (found && binding.value !== undefined) {
+          if (binding.hasLeaked) {
+            notSupportedForTransitiveMaterialization();
```

Comment:
This should be fine (at least the way things are right now), there's simply nothing TODO. Add a comment to this end.

Reply:
Done. There was a case I wanted to cover, it's not checked in as a test.